### PR TITLE
[Thumb] Resolve FIXME: We can trust live-in information from successor blocks now

### PR DESCRIPTION
**Pull request by AZero13**
As the years have gone by, live-in information has been more and more consistent and reliable, especially with changing how these are calculated.

### CODE REVIEW

File: llvm/lib/Target/ARM/ARMLoadStoreOptimizer.cpp [hunk 1]
```diff
@@ -566,18 +566,19 @@ void ARMLoadStoreOpt::UpdateBaseRegUses(MachineBasicBlock &MBB,
...
-    // information and *always* have to reset at the end of a block.
-    // See PR21029.
-    if (MBBI != MBB.end()) --MBBI;
-    BuildMI(MBB, MBBI, DL, TII->get(ARM::tSUBi8), Base)
-        .add(t1CondCodeOp(true))
-        .addReg(Base)
-        .addImm(WordOffset * 4)
-        .addImm(Pred)
-        .addReg(PredReg);
+  // Reset if Base is in the sucessor blocks' live-in sets.
```
Comment (topperc) [inline llvm/lib/Target/ARM/ARMLoadStoreOptimizer.cpp:569]: successor*

**Comment by topperc** [comment]
The PR21403 mentioned in the comment still seems to reproduce. It says R6 is missing from the liveins lists. R6 in those cases is reserved and being used as a base pointer. Reserved registers probably aren't required to be in the liveins list. Not sure if R6 is relevant to the code being changed here.

**Comment by AtariDreams** [comment]
> The PR21403 mentioned in the comment still seems to reproduce. It says R6 is missing from the liveins lists. R6 in those cases is reserved and being used as a base pointer. Reserved registers probably aren't required to be in the liveins list. Not sure if R6 is relevant to the code being changed here.

I think you have the wrong PR number

**Comment by AtariDreams** [comment]
> The PR21403 mentioned in the comment still seems to reproduce. It says R6 is missing from the liveins lists. R6 in those cases is reserved and being used as a base pointer. Reserved registers probably aren't required to be in the liveins list. Not sure if R6 is relevant to the code being changed here.

Can you tell me the case this happens in? I cannot find it.

**Comment by topperc** [comment]
> > The PR21403 mentioned in the comment still seems to reproduce. It says R6 is missing from the liveins lists. R6 in those cases is reserved and being used as a base pointer. Reserved registers probably aren't required to be in the liveins list. Not sure if R6 is relevant to the code being changed here.
> 
> Can you tell me the case this happens in? I cannot find it.

Sorry 21403 is the PR number in github. The number in the comment is the bugzilla number. https://github.com/llvm/llvm-project/issues/21403

**Comment by AtariDreams** [comment]
> > > The PR21403 mentioned in the comment still seems to reproduce. It says R6 is missing from the liveins lists. R6 in those cases is reserved and being used as a base pointer. Reserved registers probably aren't required to be in the liveins list. Not sure if R6 is relevant to the code being changed here.
> > 
> > 
> > Can you tell me the case this happens in? I cannot find it.
> 
> Sorry 21403 is the PR number in github. The number in the comment is the bugzilla number. #21403

So I used the .ll file used to repro the issue, converted it to modern synax via only addressing the errors llc gave me which oonly consisted of removing the metadata keywords themselves (but not the actual metadata) and the getelementptr/load syntax.

I compared the output before and after the patch. No difference.

**Comment by AtariDreams** [comment]
Here is the file I tested:
[oggenc.txt](https://github.com/llvm/llvm-project/files/14894270/oggenc.txt)

**Comment by topperc** [comment]
> > > > The PR21403 mentioned in the comment still seems to reproduce. It says R6 is missing from the liveins lists. R6 in those cases is reserved and being used as a base pointer. Reserved registers probably aren't required to be in the liveins list. Not sure if R6 is relevant to the code being changed here.
> > > 
> > > 
> > > Can you tell me the case this happens in? I cannot find it.
> > 
> > 
> > Sorry 21403 is the PR number in github. The number in the comment is the bugzilla number. #21403
> 
> So I used the .ll file used to repro the issue, converted it to modern synax via only addressing the errors llc gave me which oonly consisted of removing the metadata keywords themselves (but not the actual metadata) and the getelementptr/load syntax.
> 
> I compared the output before and after the patch. No difference.

Agreed that test would not be affected by this patch.

The title of this patch is "We can trust live-in information from successor blocks now" implying something is different today than it was when that FIXME was written. The test file from https://github.com/llvm/llvm-project/issues/21403 still has the same live-in behavior today as it did when the bug was filed.

Was that livein issue misinterpreted and was never really an issue?

**Comment by AtariDreams** [comment]
Also by the thing that changed I was actually referring to the changes that resulted from moving from livephysregs to liveregunits. 

R6 is reserved anyway, so it is always "live"

**Comment by topperc** [comment]
> Also by the thing that changed I was actually referring to the changes that resulted from moving from livephysregs to liveregunits.
> 
> R6 is reserved anyway, so it is always "live"

Then you should explain that in your description. I was going off the text of the FIXME.

**Comment by AtariDreams** [comment]
> > Also by the thing that changed I was actually referring to the changes that resulted from moving from livephysregs to liveregunits.
> > R6 is reserved anyway, so it is always "live"
> 
> Then you should explain that in your description. I was going off the text of the FIXME.

I think I found why some are missing from the live-in set, and this patch addresses it: https://github.com/llvm/llvm-project/pull/88023

**Pull request closed without merging**